### PR TITLE
Improve appearance of options inside the preferences dialog

### LINKED ISSUE #2481: Align boxes for different lines in preferences dialog
The following image displays the look of the preferences dialog (master). The idea is to use a QGridLayout to align for example, the first three options that include comboboxes.

image

### PR DESCRIPTION
**Pull request by goanpeca**
Fixes #2481
## Description

The comboboxes in the interface section have been aligned as well as the spinboxes in the status section of the main config page, Editor page and Ipython page.

### Main
image

### Editor
image

image

image

## IPython
image

image

## History
image

## Todo
- [x] Fix main preferences page
- Editor
    - [x] Source
    - [x] Code
    - [x] Advanced 
- Ipython
    - [x] Graphics
    - [x] Advanced settings
- [x] History

## Future PR
- Add a richtext checkbox widget.
- Improve docstrings on helper methods (currently not very helpful)

### CODE REVIEW

**Comment by goanpeca** [comment]
@Nodd due to way the preferences dialog is designed, it is not possible to use a QGridLayout or a QFormLayout. There are helpers methods that create compound widgets already, so I had to build around that.

**Comment by blink1073** [comment]
That is quite pretty!

**Comment by goanpeca** [comment]
Glad you like it hun :-)

**Comment by Nodd** [comment]
The interface looks nice but not using layouts for this looks wrong. Isn't it possible to access the undelying label and widget as `widget.label` or something ? This way you could add them to a grid layout.

**Comment by goanpeca** [comment]
Well... if you take a look at spyder/plugins/configdialog.py

**Comment by goanpeca** [comment]
Hmmm let me think, actually yes there is a way I will come in a sec with the new fix

Good idea @Nodd !

**Comment by goanpeca** [comment]
Ok @Nodd, it did made more sense like that! :)

Now....

image

:-)

**Comment by goanpeca** [comment]
If there is any other place in the preferences dialog where something might look funky, this would be the time to spill your beans!

**Comment by goanpeca** [comment]
@SylvainCorlay I think we can use some color in those QtAwesome icons ;-)

**Comment by Nodd** [comment]
* In Editor > Advanced options, the characters and width of indentation could be aligned. (maybe in the source code of the Display tab also, but the options are further apart).
* In IPython console > Graphics, the backend part
* In IPython console > Advanced, the command prompts
And that's it !

File: spyderlib/plugins/configdialog.py [hunk 1]
```diff
@@ -761,21 +769,25 @@ def setup_page(self):
         margins_layout.addWidget(margin_spin)
         prompt_box = newcb(_("Prompt when exiting"), 'prompt_on_exit')
 
-        # Language chooser
-        choices = sorted([(val, key) for key, val in LANGUAGE_CODES.items()])
-        language_combo = self.create_combobox(_('Language'), choices,
-                                              'interface_language',
-                                              restart=True)
-
         # Decide if it's possible to activate or not singie instance mode
```
Comment (Nodd) [inline spyderlib/plugins/configdialog.py]: There is a typo in `singie` int the comment.

File: spyderlib/plugins/configdialog.py [hunk 2]
```diff
@@ -761,21 +769,25 @@ def setup_page(self):
...
+        comboboxes_layout = QHBoxLayout()
+        cbs_layout = QGridLayout()
+        cbs_layout.addWidget(style_combo.label, 0, 0)
+        cbs_layout.addWidget(style_combo.combobox, 0, 1)
+        cbs_layout.addWidget(icons_combo.label, 1, 0)
+        cbs_layout.addWidget(icons_combo.combobox, 1, 1)
+        cbs_layout.addWidget(language_combo.label, 2, 0)
+        cbs_layout.addWidget(language_combo.combobox, 2, 1)
+        comboboxes_layout.addLayout(cbs_layout)
+        comboboxes_layout.addStretch(1)
```
Comment (Nodd) [inline spyderlib/plugins/configdialog.py]: What is the `1` for ? I think it's for the orientation, can't you use a `qt.something` constant instead ?

Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: http://doc.qt.io/qt-4.8/qboxlayout.html#addStretch

File: spyderlib/plugins/configdialog.py (hunk 1 above)
Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: hehe sure :-)

File: spyderlib/plugins/configdialog.py (hunk 2 above)
Comment (Nodd) [inline spyderlib/plugins/configdialog.py]: Oh, right. Nevermind.

**Comment by goanpeca** [comment]
@Nodd done :-)

Check the description for the images.

**Comment by ccordoba12** [comment]
Wow, it really looks much better!

**Comment by goanpeca** [comment]
:-) yes we are going pro now ;)

**Comment by goanpeca** [comment]
Any more comments are welcome, so I can tackle any other weirdness in the preferences

**Comment by goanpeca** [comment]
[Is there anybody out there?](https://www.youtube.com/watch?v=aHN6AViJAvI&feature=youtu.be&t=37)

**Comment by Nodd** [comment]
Beautiful !
You want more improvements to do ? the 'for', 'if', 'def' keywords should be monospace rather than between quotes. I don't know if the text in checkboxes accepts rich text...

**Comment by Nodd** [comment]
Also, #2348.

**Comment by goanpeca** [comment]
They dont but there is some code for a richtext checkbox so it could be done

**Comment by ccordoba12** [comment]
@goanpeca, I have some suggestions of my own (now that you're working on this):

1. Please reorganize how our options look in

           Editor > Code/Introspection/Analysis > Analysis

    What I mean is this:

    1. The `pep8` option shouldn't be next to the `pyflakes` one. It should be below it instead.
    2. `Code analysis (pyflakes)` could have a better name, like `Real time code analysis`.
    3. `Tasks` could have a better name too, i.e. a more informative one, but it doesn't occur to me a better one.
    4. The label below **Analysis** (i.e. `Note: add ...`) could be removed and instead add those notes as tooltips to the `pyflakes` and `pep8` buttons.

2. The `Font Style` section in `History Log` should be the first one (instead of the second one), to make it consistent with the other sections (where `Font Style` is always the first secion).

**Comment by ccordoba12** [comment]
I forgot to mention: do you plan to finish your task number **6.**?

**Comment by goanpeca** [comment]
Hi @ccordoba12 good to hear from you :smiling_imp: . I will work on your suggestions later tonight. And yes I will add the docstrings to the helper methods, cause right now they offer no help whatsoever!

**Comment by goanpeca** [comment]
Maybe instead of tasks, we can call them, `Spyder code annotations`, or , `Code annotations` or `Source code annotations`?

**Comment by goanpeca** [comment]
I think it would be nice if by default we added something to let the user know that the option has a tip on it, cause it is not immediately discoverable.

**Comment by goanpeca** [comment]
For instance we could add a small light bulb icon to the right of options that offer a tip.

**Comment by ccordoba12** [comment]
> For instance we could add a small light bulb icon to the right of options that offer a tip.

Great idea!! A help icon next to the option would be better :-) For another PR, right?

**Comment by goanpeca** [comment]
Sure that too, and yes for another PR.

What do you think about calling tasks... Code annotations?

**Comment by goanpeca** [comment]
@ccordoba12 I will improve the docstrings in a different PR, so that we can limit this to looks.

**Comment by ccordoba12** [comment]
I'd go with `Code annotations`. It's a good choice :-)

**Comment by goanpeca** [comment]
@ccordoba12 done. I will make the additional PRs soon.

File: spyderlib/plugins/editor.py [hunk 3]
```diff
@@ -219,35 +223,33 @@ def setup_page(self):
...
-                                  "%s page.") % pep8_url)
-        analysis_label.setWordWrap(True)
+        pep_url = '<a href="http://www.python.org/dev/peps/pep-0008/">PEP8</a>'
+        pep8_label = QLabel(_("<i>(Refer to the {} page)</i>").format(pep_url))
+        pep8_label.setOpenExternalLinks(True)
         is_pyflakes = codeanalysis.is_pyflakes_installed()
         is_pep8 = codeanalysis.get_checker_executable('pep8') is not None
-        analysis_label.setEnabled(is_pyflakes or is_pep8)
-        pyflakes_box = newcb(_("Code analysis")+" (pyflakes)",
+        pyflakes_box = newcb(_("Real time code analysis")+" (pyflakes)",
```
Comment (ccordoba12) [inline spyderlib/plugins/editor.py]: Please remove `pyflakes` from this text, and use the same exact text we have in `Optional dependencies` related to `pyflakes`, i.e.

     Real-time code analysis

File: spyderlib/plugins/editor.py [hunk 4]
```diff
@@ -219,35 +223,33 @@ def setup_page(self):
...
+                            "warnings will be highlighted.</p>"
+                            "<p><u>Note</u>: add <b>analysis:ignore</b> in "
+                            "a comment to ignore code analysis "
+                            "warnings.</p>"))
         pyflakes_box.setEnabled(is_pyflakes)
         if not is_pyflakes:
             pyflakes_box.setToolTip(_("Code analysis requires pyflakes %s+") %
                                     codeanalysis.PYFLAKES_REQVER)
-        pep8_box = newcb(_("Style analysis")+' (pep8)',
+        pep8_box = newcb(_("Real time style analysis")+' (pep8)',
```
Comment (ccordoba12) [inline spyderlib/plugins/editor.py]: The same as my above comment here, i.e. remove `pep8` and use as text

    Real-time code style analysis

**Comment by ccordoba12** [comment]
Thanks a lot for this! Just two minor comments and we are ready to go :-)

**Comment by goanpeca** [comment]
Done @ccordoba12

**Comment by ccordoba12** [comment]
This needs a rebase now, sorry ;-)

**Comment by goanpeca** [comment]
@ccordoba12 done

**Comment by ccordoba12** [comment]
Thanks man! This looks so much professional now! :-)

**Comment by goanpeca** [comment]
:+1: 

@Nodd 

I have a new rich checkbox on the works, it is almost done,

image

**Comment by Nodd** [comment]
I must be more careful with the ideas I give you… I hope that the implementation is not too hacky !

**Comment by goanpeca** [comment]
Not too much

I based most of the code from a C++ code in a [qt forum](http://www.qtcentre.org/threads/60455-SOLVED-Issues-with-displaying-rich-text-for-a-QCheckBox)

I just made some tweaking ;-)

**Pull request merged**